tests/components/alarmdecoder/test_config_flow.py

```diff
@@ -0,0 +1,424 @@
+from alarmdecoder.util import NoDeviceError
+import pytest
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.components.alarmdecoder import config_flow
+from homeassistant.components.alarmdecoder.const import (
+    CONF_ALT_NIGHT_MODE,
+    CONF_AUTO_BYPASS,
+    CONF_CODE_ARM_REQUIRED,
+    CONF_DEVICE_BAUD,
+    CONF_DEVICE_PATH,
+    CONF_RELAY_ADDR,
+    CONF_RELAY_CHAN,
+    CONF_ZONE_LOOP,
+    CONF_ZONE_NAME,
+    CONF_ZONE_NUMBER,
+    CONF_ZONE_RFID,
+    CONF_ZONE_TYPE,
+    DEFAULT_ARM_OPTIONS,
+    DEFAULT_ZONE_OPTIONS,
+    DOMAIN,
+    OPTIONS_ARM,
+    OPTIONS_ZONES,
+    PROTOCOL_SERIAL,
+    PROTOCOL_SOCKET,
+)
+from homeassistant.components.binary_sensor import DEVICE_CLASS_WINDOW
+from homeassistant.const import CONF_HOST, CONF_PORT, CONF_PROTOCOL
+from homeassistant.core import HomeAssistant
+
+from tests.async_mock import patch
+from tests.common import MockConfigEntry
+
+
+@pytest.mark.parametrize(
+    "protocol,connection,baud,title",
+    [
+        (
+            PROTOCOL_SOCKET,
+            {
+                CONF_HOST: "alarmdecoder123",
+                CONF_PORT: 10001,
+            },
+            None,
+            "alarmdecoder123:10001",
+        ),
+        (
+            PROTOCOL_SERIAL,
+            {
+                CONF_DEVICE_PATH: "/dev/ttyUSB123",
+                CONF_DEVICE_BAUD: 115000,
+            },
+            115000,
+            "/dev/ttyUSB123",
+        ),
+    ],
+)
+async def test_setups(hass: HomeAssistant, protocol, connection, baud, title):
+    """Test flow for setting up the available AlarmDecoder protocols."""
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_PROTOCOL: protocol},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "protocol"
+
+    with patch("homeassistant.components.alarmdecoder.config_flow.AdExt.open"), patch(
+        "homeassistant.components.alarmdecoder.config_flow.AdExt.close"
+    ), patch(
+        "homeassistant.components.alarmdecoder.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.alarmdecoder.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], connection
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+        assert result["title"] == title
+        assert result["data"] == {
+            **connection,
+            CONF_PROTOCOL: protocol,
+            CONF_DEVICE_BAUD: baud,
+        }
+
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6158285)